src/execution_plan/execution_plan.c

```diff
@@ -498,304 +482,360 @@ static void _NewExecutionPlanSegment(RedisModuleCtx *ctx, GraphContext *gc,
+	uint *modifies = _buildModifiesArray(segment->projections);
+
+	// Our fundamental operation will be a projection or aggregation.
+	OpBase *op_project;
+	if(aggregate) {
+		op_project = NewAggregateOp(segment->projections, modifies);
+	} else {
+		op_project = NewProjectOp(segment->projections, modifies);
 	}
+	Vector_Push(ops, op_project);
 
-	bool create_clause = AST_ContainsClause(ast, CYPHER_AST_CREATE);
-	if(create_clause) {
-		AST_CreateContext create_ast_ctx = AST_PrepareCreateOp(gc, record_map, ast, qg);
-		OpBase *opCreate = NewCreateOp(stats,
-									   create_ast_ctx.nodes_to_create,
-									   create_ast_ctx.edges_to_create);
-		Vector_Push(ops, opCreate);
+	/* Add modifier operations in order such that the final execution plan will follow the sequence:
+	 * Limit -> Skip -> Sort -> Distinct -> Project/Aggregate */
+	if(distinct) {
+		OpBase *op_distinct = NewDistinctOp();
+		Vector_Push(ops, op_distinct);
 	}
 
-	const cypher_astnode_t *merge_clause = AST_GetClause(ast, CYPHER_AST_MERGE);
-	if(merge_clause) {
-		// A merge clause provides a single path that must exist or be created.
-		// As with paths in a MATCH query, build the appropriate traversal operations
-		// and append them to the set of ops.
-		AST_MergeContext merge_ast_ctx = AST_PrepareMergeOp(gc, record_map, ast, merge_clause, qg);
+	if(sort) {
+		// The sort operation will obey a specified limit, but must account for skipped records
+		uint sort_limit = (limit != RESULTSET_UNLIMITED) ? limit + skip : 0;
+		OpBase *op_sort = NewSortOp(segment->order_expressions, sort, sort_limit);
+		Vector_Push(ops, op_sort);
+	}
 
-		// Append a merge operation
-		OpBase *opMerge = NewMergeOp(stats,
-									 merge_ast_ctx.nodes_to_merge,
-									 merge_ast_ctx.edges_to_merge);
-		Vector_Push(ops, opMerge);
+	if(skip) {
+		OpBase *op_skip = NewSkipOp(skip);
+		Vector_Push(ops, op_skip);
 	}
 
-	const cypher_astnode_t *delete_clause = AST_GetClause(ast, CYPHER_AST_DELETE);
-	if(delete_clause) {
-		uint *nodes_ref;
-		uint *edges_ref;
-		AST_PrepareDeleteOp(delete_clause, qg, record_map, &nodes_ref, &edges_ref);
-		OpBase *opDelete = NewDeleteOp(nodes_ref, edges_ref, stats);
-		Vector_Push(ops, opDelete);
+	if(limit != RESULTSET_UNLIMITED) {
+		OpBase *op_limit = NewLimitOp(limit);
+		Vector_Push(ops, op_limit);
 	}
+}
 
-	const cypher_astnode_t *set_clause = AST_GetClause(ast, CYPHER_AST_SET);
-	if(set_clause) {
-		// Create a context for each update expression.
-		uint nitems;
-		EntityUpdateEvalCtx *update_exps = AST_PrepareUpdateOp(set_clause, record_map, &nitems);
-		OpBase *op_update = NewUpdateOp(gc, update_exps, nitems, stats);
-		Vector_Push(ops, op_update);
+// The RETURN logic is identical to WITH-culminating segments,
+// though a different set of API endpoints must be used for each.
+static void _buildReturnOps(ExecutionPlanSegment *segment, const cypher_astnode_t *clause,
+							Vector *ops) {
+	bool aggregate = AST_ClauseContainsAggregation(clause);
```

Comment:
In my eyes functions such as `AST_ClauseContainsAggregation` which operates on a`cypher_astnode_t`
might have been part of the actual parser, as such consider introducing a parserExt similar to GraphBLASExt where we have functions extending GraphBLAS following GraphBLAS look and feel.

`AST_ClauseContainsAggregation` -> `cypher_ast_ext_return_contains_aggregation`

Reply:
I like the parserExt idea in general, but to the parser, SUM() and ID() are both just of type `CYPHER_AST_APPLY_OPERATOR`. We register and differentiate between simple and aggregate functions in-module.